### PR TITLE
Direct frelax solve -- bugfix

### PR DESCRIPTION
**Pull request by oseikuffuor1**
Bugfix to support user-provided BoomerAMG solver for F-relaxation and also Gaussian Elimination for F-relaxation.

### CODE REVIEW

**Review by victorapm — APPROVED** [review]
Thanks!!

File: src/parcsr_ls/par_mgr_setup.c [hunk 1]
```diff
@@ -1179,78 +1177,104 @@ hypre_MGRSetup( void               *mgr_vdata,
...
-                  A_ff_array[lev] = A_ff_ptr;
+                  if (((hypre_ParAMGData*)aff_solver[lev])->A_array[0] != NULL)
+                  {
+                     // F-solver is already set up, only need to store A_ff_ptr
+                     A_ff_ptr = ((hypre_ParAMGData*)aff_solver[lev])->A_array[0];
+                     A_ff_array[lev] = A_ff_ptr;
+                  }
+                  else
+                  {
+                     if (my_id == 0)
```
Comment (rfalgout) [inline src/parcsr_ls/par_mgr_setup.c]: You should protect this with a `print_level` check or use the `hypre_error_w_msg` routine (I'm not sure I got the name exactly right).

File: src/parcsr_ls/par_mgr_setup.c [hunk 2]
```diff
@@ -1179,78 +1177,104 @@ hypre_MGRSetup( void               *mgr_vdata,
...
+                  A_ff_array[lev] = A_ff_ptr;
+                  (mgr_data -> fsolver_mode) = 1;
                }
             }
-            else // F-relaxation solver prescribed but not set up
+            else
             {
-               // Compute A_ff and setup F-solver
-               if (exec == HYPRE_EXEC_HOST)
+               if (my_id == 0)
```
Comment (rfalgout) [inline src/parcsr_ls/par_mgr_setup.c]: Same comment as above.

**Review by rfalgout — APPROVED** [review]
Looks good to me modulo my comments.  Be sure to run through valgrind since some of the memory management changed here.  Thanks!

File: src/parcsr_ls/par_mgr_setup.c [hunk 3]
```diff
@@ -1197,9 +1197,9 @@ hypre_MGRSetup( void               *mgr_vdata,
                   }
                   else
                   {
-                     if (my_id == 0)
+                     if (my_id == 0 && print_level > 1)
                      {
-                        hypre_printf("Error!!! Invalid AMG setup for user-prescribed F-relaxation.\n");
+                        hypre_error_w_msg(HYPRE_ERROR_GENERIC, "Error!!! Invalid AMG setup for user-prescribed F-relaxation.\n");
                         hypre_error(1);
```
Comment (rfalgout) [inline src/parcsr_ls/par_mgr_setup.c]: You can delete this line BTW since you've already signaled the error above.

Comment (oseikuffuor1) [reply src/parcsr_ls/par_mgr_setup.c]: @rfalgout just confirming you mean the hypre_error(1); line?

Comment (rfalgout) [reply src/parcsr_ls/par_mgr_setup.c]: Yes.  Sorry for not being clear.

Comment (oseikuffuor1) [reply src/parcsr_ls/par_mgr_setup.c]: Got it. Thanks!

**Pull request merged**